BUG: fix tz-aware DatetimeIndex + TimedeltaIndex (#17558)

azjps opened the pull request:
Fix a bug that was causing a tz-aware DatetimeIndex + TimedeltaIndex to raise an error. (An internal function was documented to return an integer view, but instead was trying to use the DatetimeIndex's dtype.) Fix another bug causing the sum of a tz-aware DatetimeIndex and a numpy array of timedeltas to incorrectly have timezone applied twice, by adding another case in `DatetimeIndex.__add__`.

- [x] closes #17558 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

azjps committed:
BUG: fix tz-aware DatetimeIndex + TimedeltaIndex (#17558)

Fix minor bug causing DatetimeIndex + TimedeltaIndex to raise an error,
and fix another bug causing the sum of a tz-aware DatetimeIndex and a
numpy array of timedeltas to incorrectly have timezone applied twice.

gfyoung commented on pandas/tests/indexes/datetimes/test_ops.py:
```diff
@@ -460,6 +460,20 @@ def test_add_dti_dti(self):
+        # and DatetimeIndex + TimedeltaIndex work as expected
+        dti = pd.DatetimeIndex([pd.Timestamp("2017/01/01")])
+        dti = dti.tz_localize('US/Eastern')
+        expected = pd.DatetimeIndex([pd.Timestamp("2017/01/01 01:00")])
```
Nit : add newline above this one.

gfyoung commented on pandas/tests/indexes/datetimes/test_ops.py:
```diff
@@ -460,6 +460,20 @@ def test_add_dti_dti(self):
+        td_np = np.array([np.timedelta64(1, 'h')], dtype="timedelta64[ns]")
+        tm.assert_index_equal(dti + td_np, expected)
+        tm.assert_index_equal(dti + td_np[0], expected)
+        tm.assert_index_equal(dti + TimedeltaIndex(td_np), expected)
```
Can we just for-loop these checks instead?  Also, let's be a little more explicit and write it this way:
~~~python
actual = dti + ...
tm.assert_index_equal(actual, expected)
~~~

gfyoung commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -651,6 +652,15 @@ def __add__(self, other):
+                if is_timedelta64_dtype(other):
+                    return self._add_delta(TimedeltaIndex(other))
+                elif is_integer_dtype(other):
+                    return NotImplemented
```
Do we have a test that hits this branch?

azjps replied:
Yes but indirectly, without this branch it was breaking a couple of other tests that added a `PeriodIndex` and a `np.array(dtype=int64)`. I didn't check what the behavior is there, but could add a more direct test.

gfyoung replied:
Direct test is certainly appreciated.  Go for it!

azjps replied:
Okay, I looked into the cases getting triggered with `PeriodIndex` and unfortunately they're buried pretty deeply in internals (e.g. not obvious this would get called from any public API, at least any more so than the tests that were already reaching this). Also @jreback commented that this should only be internal facing, so I figure its better not to "advertise" this API in a test.

gfyoung commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -651,6 +652,15 @@ def __add__(self, other):
+                elif is_integer_dtype(other):
+                    return NotImplemented
+                else:
+                    raise TypeError("cannot add {typ1} and np.ndarray[{typ2}]"
```
Same question as with the `NotImplemented`.  Let's make sure we have a test to hit this branch.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -506,6 +506,7 @@ Indexing
 - Bug in ``CategoricalIndex`` reindexing in which specified indices containing duplicates were not being respected (:issue:`17323`)
 - Bug in intersection of ``RangeIndex`` with negative step (:issue:`17296`)
 - Bug in ``IntervalIndex`` where performing a scalar lookup fails for included right endpoints of non-overlapping monotonic decreasing indexes (:issue:`16417`, :issue:`17271`)
+- Bug in adding ``DatetimeIndex`` with a ``TimedeltaIndex`` or a numpy array with ``dtype="timedelta64"`` (:issue:`17558`)
```
this is not correct; the bug is *only* with a numpy array of ``dtype=timedelta64[ns]`` (timedelta64 is not a valid dtype for pandas)

azjps replied:
The "or" was a bit confusing, but there are two separate bugs:
* adding tz-aware ``DatetimeIndex`` with ``TimedeltaIndex``
* adding tz-aware ``DatetimeIndex`` with numpy array which is timedelta64-dtype-like

jreback replied:
how so? where does DTI + TDI tz-aware not work? your issues is w.r.t. np.array of timedeltas

azjps replied:
I mentioned it briefly in the issue ticket, although I neglected to post the full stack trace:

```python
In [1]: import numpy as np, pandas as pd

In [2]: dti = pd.DatetimeIndex([pd.Timestamp("2017/01/01")]).tz_localize('US/Eastern')

In [3]: td_np = np.array([np.timedelta64(1, 'ns')]) 

In [4]: (dti + pd.TimedeltaIndex(td_np))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-20f28143c2bb> in <module>()
----> 1 (dti + pd.TimedeltaIndex(td_np))

/spare/local/azhu/venv-el6-20170811-stonelib/lib/python2.7/site-packages/pandas/tseries/base.pyc in __add__(self, other)
    617             from pandas.tseries.offsets import DateOffset
    618             if isinstance(other, TimedeltaIndex):
--> 619                 return self._add_delta(other)
    620             elif isinstance(self, TimedeltaIndex) and isinstance(other, Index):
    621                 if hasattr(other, '_add_delta'):

/spare/local/azhu/venv-el6-20170811-stonelib/lib/python2.7/site-packages/pandas/tseries/index.pyc in _add_delta(self, delta)
    787             new_values = self._add_delta_td(delta)
    788         elif isinstance(delta, TimedeltaIndex):
--> 789             new_values = self._add_delta_tdi(delta)
    790             # update name when delta is Index
    791             name = com._maybe_match_name(self, delta)

/spare/local/azhu/venv-el6-20170811-stonelib/lib/python2.7/site-packages/pandas/tseries/base.pyc in _add_delta_tdi(self, other)
    704             mask = (self._isnan) | (other._isnan)
    705             new_values[mask] = tslib.iNaT
--> 706         return new_values.view(self.dtype)
    707 
    708     def isin(self, values):

TypeError: data type not understood
```

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -651,6 +652,15 @@ def __add__(self, other):
                                         typ2=type(other).__name__))
+            elif isinstance(other, np.ndarray):
+                if is_timedelta64_dtype(other):
+                    return self._add_delta(TimedeltaIndex(other))
```
you can only accept timedelta64 here, integer is not valid either fro datetimelikes in add

azjps replied:
As noted in another comment above, the integer case is being used by internals for ``PeriodIndex``. Example: [`pandas/tseries/offsets.py:_beg_apply_index#L468`](https://github.com/pandas-dev/pandas/blob/master/pandas/tseries/offsets.py#L468) adds `base_period (PeriodIndex) + roll (ndarray[dtype=int64])`.

jreback replied:
no we don't allow this user facting; this is ONLY an internal operation

azjps replied:
Sure, but I don't quite understand what you're getting at -- are you saying there's still something to be done here? I just added the integer branch to prevent aforementioned internally used operation from breaking. (since it returns `NotImplemented` as it currently does, it will fall through to ufunc implementation.) I can add a code comment to explain the intent?

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -731,7 +741,7 @@ def _add_delta_tdi(self, other):
             mask = (self._isnan) | (other._isnan)
             new_values[mask] = iNaT
-        return new_values.view(self.dtype)
+        return new_values.view('i8')
```
leave this

azjps replied:
This fixes adding tz-aware `DatetimeIndex` and `TimedeltaIndex`, since `self.dtype` here is a `Datetime64TZDtype` which causes `new_values.view(..)` to raise an exception. Would you prefer another solution, say checking if `is_datetime64tz_dtype(self.dtype)` first?

jreback replied:
yes use is_datetime64tz_dtype(self.dtype) is the correct idiom. but you'll have to demonstrate this bug specifically

azjps replied:
Posted stack trace that hits this line above. So I can change this to
```python
if is_datetime64tz_dtype(self.dtype):
  return new_values.view('i8')
return new_values.view(self.dtype)
```

jreback replied:
this is the not the right fix.

tz-aware need to be turned into local i8, added, then turned back into the original timezone via localization. normally we *do* add i8 directly, BUT you cannot do this with tz-aware, e.g. the days are shifted otherwise (imagine adding days), or don't work when you cross DST.

pandas.tseries.offsets  already does all of this. I haven't stepped thru this code lately, but need to hit a similar path to that.

jreback commented on pandas/tests/indexes/datetimes/test_ops.py:
```diff
@@ -460,6 +460,20 @@ def test_add_dti_dti(self):
         with pytest.raises(TypeError):
             dti + dti_tz
 
+    def test_add_dti_ndarray(self):
```
add a test which raises for numpy arrays of incorrect type (e.g. not m8[ns]);

jreback commented on pandas/tests/indexes/datetimes/test_ops.py:
```diff
@@ -460,6 +460,20 @@ def test_add_dti_dti(self):
+        # GH 17558
+        # Check that tz-aware DatetimeIndex + np.array(dtype="timedelta64")
+        # and DatetimeIndex + TimedeltaIndex work as expected
+        dti = pd.DatetimeIndex([pd.Timestamp("2017/01/01")])
```
you can move this entire test to ``datetimelike.py``; it will hit datetime, period, timedelta that way

jreback replied:
we should prob move more of these add tests there as well, but can do that in another PR

azjps committed:
Add DatetimeIndex and timedeltas via _add_delta_td

The internal implementation of _add_delta_td already correctly
adds a datetime index and an array of timedeltas, so use that
implementation instead of wrapping in a TimedeltaIndex (which
requires a bit more metadata like name to be passed around).

Move test to datetimelike to check addition of TimedeltaIndex
and np.array(timedelta64) with each of {DatetimeIndex,
TimedeltaIndex, PeriodIndex}. Fix the latter two to explicitly
support addition with a numpy array.

Clarify some comments such as the whatsnew and move tests to
test_datetimelike.

pep8speaks commented:
Hello @azjps! Thanks for updating the PR.









 - In the file [`pandas/tests/indexes/datetimelike.py`](https://github.com/pandas-dev/pandas/blob/c9da215ba5b9b8220349f033fd0241f973cbf961/pandas/tests/indexes/datetimelike.py), following are the PEP8 issues :

> [Line 51:11](https://github.com/pandas-dev/pandas/blob/c9da215ba5b9b8220349f033fd0241f973cbf961/pandas/tests/indexes/datetimelike.py#L51): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 70:1](https://github.com/pandas-dev/pandas/blob/c9da215ba5b9b8220349f033fd0241f973cbf961/pandas/tests/indexes/datetimelike.py#L70): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

azjps commented on pandas/tests/indexes/datetimelike.py:
```diff
@@ -38,3 +40,31 @@ def test_view(self):
+          idx = idx.tz_localize("US/Eastern")
+
+        expected = idx + np.timedelta64(1, 'D')
+        tm.assert_index_equal(idx, expected - np.timedelta64(1, 'D'))
```
just to explain the awkwardness here: was trying to figure out how to demonstrate that `expected` actually has the properties we expect, e.g. has `name = "x"` and is a day ahead of `idx`

azjps commented on pandas/tests/indexes/datetimes/test_datetimelike.py:
```diff
@@ -76,3 +76,27 @@ def test_union(self):
             result = first.union(case)
             assert tm.equalContents(result, everything)
+
+    def test_add_dti_td(self):
```
this is almost the same as special case of the above test in `datetimelike.py` for `DatetimeIndex`, if that test looks good I will remove this one.

jreback replied:
see my comments above you can simply re-define create_index to make this work (in another test class)

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -651,6 +652,18 @@ def __add__(self, other):
+            elif isinstance(other, np.ndarray):
+                if is_timedelta64_dtype(other):
+                    return self._add_delta(other)
+                elif is_integer_dtype(other):
```
I think instead of this, we want to specifically define __add__ in PeriodIndex which handles this case and calls super otherwise.

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -731,6 +744,8 @@ def _add_delta_tdi(self, other):
             mask = (self._isnan) | (other._isnan)
             new_values[mask] = iNaT
+        if is_datetime64tz_dtype(self.dtype):
+            return new_values.view('i8')
```
no, see my comments

jreback commented on pandas/core/indexes/datetimes.py:
```diff
@@ -801,6 +802,8 @@ def _add_delta(self, delta):
 
         if isinstance(delta, (Tick, timedelta, np.timedelta64)):
             new_values = self._add_delta_td(delta)
+        elif isinstance(delta, np.ndarray) and is_timedelta64_dtype(delta):
```
this should simply be incorporated into the case right below

azjps replied:
by below, do I understand you correctly that we can consolidate as:

```python
if isinstance(delta, (Tick, timedelta, np.timedelta64, np.ndarray)):
    new_values = self._add_delta_td(delta)
elif ..:
```

or are you suggesting to wrap `delta` into a `TimedeltaIndex`?

jreback replied:
yes
but i think we need to check the dtype as well if it's a numpy array (can be separate from this)

jreback commented on pandas/core/indexes/timedeltas.py:
```diff
@@ -312,9 +312,11 @@ def _maybe_update_attributes(self, attrs):
         if isinstance(delta, (Tick, timedelta, np.timedelta64)):
             new_values = self._add_delta_td(delta)
-            name = self.name
+        elif isinstance(delta, np.ndarray) and is_timedelta64_dtype(delta):
```
same comment

jreback commented on pandas/tests/indexes/datetimelike.py:
```diff
@@ -38,3 +41,30 @@ def test_view(self):
+        # and DatetimeIndex + TimedeltaIndex work as expected
+        idx = self.create_index()
+        idx.name = "x"
+        if isinstance(idx, pd.DatetimeIndex):
```
don't use an if like this. 

let's create a sub-class that specifically tests tz-aware (so .create_index()) works correctly.  (add in datetimes/test_datetimelikes)

azjps replied:
Totally agree with this, but unfortunately a tz-aware-based subclass fails a whole bunch of other common tests, so the scope of this PR would have to be expanded further (for which I'd probably need help). Here's one example to illustrate, basically `DatetimeIndex` constructor will read its input list into a (tz-naive) numpy array and "re-tz_localize" based on dtype's timezone:

```python
    def test_repr_roundtrip(self):
    
        idx = self.create_index()
>       tm.assert_index_equal(eval(repr(idx)), idx)

>       raise AssertionError(msg)
E       AssertionError: Index are different
E       
E       Index values are different (100.0 %)
E       [left]:  DatetimeIndex(['2013-01-01 05:00:00-05:00', '2013-01-02 05:00:00-05:00',
E                      '2013-01-03 05:00:00-05:00', '2013-01-04 05:00:00-05:00',
E                      '2013-01-05 05:00:00-05:00'],
E                     dtype='datetime64[ns, US/Eastern]', freq='D')
E       [right]: DatetimeIndex(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00',
E                      '2013-01-03 00:00:00-05:00', '2013-01-04 00:00:00-05:00',
E                      '2013-01-05 00:00:00-05:00'],
E                     dtype='datetime64[ns, US/Eastern]', freq='D')

(Pdb) repr(idx)  # broke up output for readability
"DatetimeIndex(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00',\n"
"               '2013-01-03 00:00:00-05:00', '2013-01-04 00:00:00-05:00',\n"
"               '2013-01-05 00:00:00-05:00'],\n"
"              dtype='datetime64[ns, US/Eastern]', freq='D')"
(Pdb) eval(repr(idx))
DatetimeIndex(['2013-01-01 05:00:00-05:00', '2013-01-02 05:00:00-05:00',
               '2013-01-03 05:00:00-05:00', '2013-01-04 05:00:00-05:00',
               '2013-01-05 05:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq=None)
```
Most of the other test failures can also be treated as variants of this, as these tests also construct a separate `DatetimeIndex` instance without tz information, although its possible to (say) modify those tests directly to `tz_localize` to UTC and then `tz_convert` to appropriate timezone if `create_index().tz` is available.

Also the tests related to union, intersection, symmetric_difference, etc try to do these operations against the underlying numpy `\.values` which are tz-naive, leading to issues. We maybe could get around these by just disabling that part of the test.

jreback replied:
So I think the right way to do this is to do another PR which separates out the DTI tests to 2 classes for tz-naive and tz-aware and modifies create_index as appropriate. If there are failing tests we can look closer. Want to start on that?

jreback reviewed (changes requested):
put on hold until testing is unwound.

jreback commented:
can you rebase and move the note to 0.21.1

jreback commented:
can you rebase / update. we have slightly better testing in place now FYI.

jreback commented:
closing as stale
